File: apps/ios/Packages/StoryArcKit/Sources/Formats/ImageFolderArchive.swift
public import Foundation

/// A directory of ordered images, read as one publication.
///
/// `publication-formats` lists a plain folder alongside the archive formats, and
/// it behaves like one: the same page filter, the same natural sort, the same
/// `ComicInfo.xml` handling. The only difference is that there is no container to
/// parse, so this is the one reader with no format work in it at all.
///
/// Subdirectories are walked, because chapters-as-folders is how unpacked comics
/// are actually laid out — and ordering by full path makes `ch10` follow `ch2`
/// for a folder exactly as it does inside a CBZ.
public struct ImageFolderArchive: ComicArchiveReading {
    public let pages: [PageEntry]
    public let skippedPageCount: Int
    /// `ComicInfo.xml` contents when the folder carries one.
    public let comicInfoData: Data?

    private let root: URL

    public init(directory: URL) throws {
        var isDirectory: ObjCBool = false
        guard FileManager.default.fileExists(atPath: directory.path, isDirectory: &isDirectory),
              isDirectory.boolValue
        else { throw ComicArchiveError.unrecognisedContainer }

        self.root = directory.standardizedFileURL

        // Symbolic links are not followed. A publication folder is chosen by the
        // user, but it is still untrusted input: a link pointing outside the root
        // would let a crafted folder read arbitrary files, and no real comic
        // needs one.
        let walker = FileManager.default.enumerator(
            at: root,
            includingPropertiesForKeys: [.isRegularFileKey, .fileSizeKey, .isSymbolicLinkKey],
            options: [.skipsHiddenFiles, .producesRelativePathURLs]
        )

        var candidates: [PageEntry] = []
        var skipped = 0
        var comicInfo: URL?

        while let entry = walker?.nextObject() as? URL {
            let values = try? entry.resourceValues(
                forKeys: [.isRegularFileKey, .fileSizeKey, .isSymbolicLinkKey]
            )
            guard values?.isSymbolicLink != true, values?.isRegularFile == true else { continue }

            let relative = Self.relativePath(of: entry, under: root)
            if relative.lowercased().hasSuffix("comicinfo.xml") {
                comicInfo = entry
                continue
            }
            guard PageOrdering.isPage(path: relative) else { continue }

            let size = values?.fileSize ?? 0
            // A zero-length file is a page that will never decode. Counting it as
            // skipped is what lets the reader say "opened 10, skipped 2".
            if size == 0 {
                skipped += 1
                continue
            }
            candidates.append(PageEntry(path: relative, byteCount: size))
        }

        self.pages = PageOrdering.sorted(candidates)
        self.skippedPageCount = skipped
        self.comicInfoData = comicInfo.flatMap { try? Data(contentsOf: $0) }
    }

    public func data(for page: PageEntry) async throws -> Data {
        let url = root.appending(path: page.path).standardizedFileURL
        // Re-checked at read time rather than trusted from the walk: the path
        // came off a filesystem that can change between indexing and reading.
        guard url.path.hasPrefix(root.path) else { throw ComicArchiveError.unreadable }
        guard let data = try? Data(contentsOf: url) else { throw ComicArchiveError.unreadable }
        return data
    }

    private static func relativePath(of url: URL, under root: URL) -> String {
        let full = url.standardizedFileURL.path
        let base = root.path.hasSuffix("/") ? root.path : root.path + "/"
        return full.hasPrefix(base) ? String(full.dropFirst(base.count)) : url.lastPathComponent
    }
}

File: apps/android/core/format/src/main/kotlin/app/storyarc/core/format/ImageFolderArchive.kt
package app.storyarc.core.format

import java.io.File

/**
 * A directory of ordered images, read as one publication.
 *
 * `publication-formats` lists a plain folder alongside the archive formats, and it
 * behaves like one: the same page filter, the same natural sort, the same
 * `ComicInfo.xml` handling. The only difference is that there is no container to
 * parse, so this is the one reader with no format work in it at all.
 *
 * Subdirectories are walked, because chapters-as-folders is how unpacked comics
 * are actually laid out — and ordering by full path makes `ch10` follow `ch2` for
 * a folder exactly as it does inside a CBZ.
 */
class ImageFolderArchive private constructor(
    private val root: File,
    override val pages: List<PageEntry>,
    override val skippedPageCount: Int,
    /** `ComicInfo.xml` contents when the folder carries one. */
    val comicInfoData: ByteArray?,
) : ComicArchiveReading {

    /** The folder's parsed metadata, when it carries any. */
    val comicInfo: ComicInfo? by lazy { comicInfoData?.let(ComicInfo::parse) }

    override val coverPage: PageEntry?
        get() = CoverSelection.cover(pages, comicInfo?.coverPageIndex)

    override val doublePageIndices: List<Int>
        get() = PageDeclarations.spreads(pages, comicInfo?.doublePageIndices.orEmpty())


    companion object {
        fun open(directory: File): ImageFolderArchive {
            if (!directory.isDirectory) throw ComicArchiveException.UnrecognisedContainer()
            val root = directory.canonicalFile

            val candidates = mutableListOf<PageEntry>()
            var skipped = 0
            var comicInfo: File? = null

            for (file in root.walkTopDown()) {
                if (!file.isFile) continue
                // Symbolic links are not followed. A publication folder is chosen
                // by the user, but it is still untrusted input: a link pointing
                // outside the root would let a crafted folder read arbitrary
                // files, and no real comic needs one.
                if (file.canonicalFile != file.absoluteFile) continue

                val relative = file.relativeTo(root).invariantSeparatorsPath
                if (relative.lowercase().endsWith("comicinfo.xml")) {
                    comicInfo = file
                    continue
                }
                if (!PageOrdering.isPage(relative)) continue

                // A zero-length file is a page that will never decode. Counting it
                // as skipped is what lets the reader say "opened 10, skipped 2".
                if (file.length() == 0L) {
                    skipped++
                    continue
                }
                candidates += PageEntry(relative, file.length())
            }

            return ImageFolderArchive(
                root = root,
                pages = PageOrdering.sorted(candidates),
                skippedPageCount = skipped,
                comicInfoData = comicInfo?.readBytes(),
            )
        }
    }

    override suspend fun data(page: PageEntry): ByteArray {
        val file = File(root, page.path).canonicalFile
        // Re-checked at read time rather than trusted from the walk: the path came
        // off a filesystem that can change between indexing and reading.
        if (!file.path.startsWith(root.path + File.separator)) {
            throw ComicArchiveException.Unreadable()
        }
        if (!file.isFile) throw ComicArchiveException.Unreadable()
        return file.readBytes()
    }
}
